crates/sui-types/src/passkey_authenticator.rs

```diff
@@ -0,0 +1,309 @@
+    fn verify_user_authenticator_epoch(
+        &self,
+        _epoch: EpochId,
+        _max_epoch_upper_bound_delta: Option<u64>,
+    ) -> SuiResult {
+        Ok(())
+    }
+
+    /// Verify an intent message of a transaction with an passkey authenticator.
+    fn verify_claims<T>(
+        &self,
+        intent_msg: &IntentMessage<T>,
+        author: SuiAddress,
+        _aux_verify_data: &VerifyParams,
+        _zklogin_inputs_cache: Arc<VerifiedDigestCache<ZKLoginInputsDigest>>,
+    ) -> SuiResult
+    where
+        T: Serialize,
+    {
+        // Check the intent and tx_digest is consisted from what's parsed from client_data_json.challenge
+        if intent_msg.intent != self.intent || to_tx_digest(intent_msg) != self.tx_digest {
+            return Err(SuiError::InvalidSignature {
+                error: "Invalid challenge".to_string(),
+            });
+        };
+
+        // Construct msg = authenticator_data || sha256(client_data_json).
+        let mut message = self.authenticator_data.clone();
+        let client_data_hash = Sha256::digest(self.client_data_json.as_slice()).digest;
+        message.extend_from_slice(&client_data_hash);
+
+        // Check if author is derived from the public key.
+        if author != SuiAddress::from(&self.get_pk()?) {
+            return Err(SuiError::InvalidSignature {
+                error: "Invalid author".to_string(),
+            });
+        };
+
+        // Verify the signature against pk and message.
+        self.pk
+            .verify(&message, &self.signature)
+            .map_err(|_| SuiError::InvalidSignature {
+                error: "Fails to verify".to_string(),
+            })
+    }
+}
+
+impl ToFromBytes for PasskeyAuthenticator {
+    fn from_bytes(bytes: &[u8]) -> Result<Self, FastCryptoError> {
+        // The first byte matches the flag of PasskeyAuthenticator.
+        if bytes.first().ok_or(FastCryptoError::InvalidInput)?
+            != &SignatureScheme::PasskeyAuthenticator.flag()
+        {
+            return Err(FastCryptoError::InvalidInput);
+        }
+        let passkey: PasskeyAuthenticator =
+            bcs::from_bytes(&bytes[1..]).map_err(|_| FastCryptoError::InvalidSignature)?;
+        Ok(passkey)
+    }
+}
+
+impl AsRef<[u8]> for PasskeyAuthenticator {
+    fn as_ref(&self) -> &[u8] {
+        self.bytes
+            .get_or_try_init::<_, eyre::Report>(|| {
+                let as_bytes = bcs::to_bytes(self).expect("BCS serialization should not fail");
+                let mut bytes = Vec::with_capacity(1 + as_bytes.len());
+                bytes.push(SignatureScheme::PasskeyAuthenticator.flag());
+                bytes.extend_from_slice(as_bytes.as_slice());
+                Ok(bytes)
+            })
+            .expect("OnceCell invariant violated")
+    }
+}
+/// Compute the digest that the signature committed over as `intent || hash(tx_data)`, total
+/// of 3 + 32 = 35 bytes.
+pub fn to_signing_digest<T: Serialize>(
```

Comment:
if we don't hash it again outside the signing alg, let's call it signing message or input